src/lobpcg/svd.rs

```diff
@@ -0,0 +1,327 @@
+                .for_each(|(mut a, b)| a.mapv_inplace(|x| x / *b));
+
+            (ularge, vlarge)
+        } else {
+            let ularge = self.eigvecs.select(Axis(1), &indices);
+
+            let mut vlarge = self.problem.t().dot(&ularge);
+            vlarge
+                .columns_mut()
+                .into_iter()
+                .zip(values.iter())
+                .for_each(|(mut a, b)| a.mapv_inplace(|x| x / *b));
+
+            (ularge, vlarge)
+        };
+
+        (u, values, v.reversed_axes())
+    }
+}
+
+/// Truncated singular value decomposition
+///
+/// Wraps the LOBPCG algorithm and provides convenient builder-pattern access to
+/// parameter like maximal iteration, precision and constrain matrix.
+pub struct TruncatedSvd<A: NdFloat> {
+    order: Order,
+    problem: Array2<A>,
+    precision: f32,
+    maxiter: usize,
+}
+
+impl<A: Float + NdFloat + PartialOrd + Default + Sum> TruncatedSvd<A> {
+    /// Create a new truncated SVD problem
+    ///
+    /// # Parameters
+    ///  * `problem`: rectangular matrix which is decomposed
+    ///  * `order`: whether to return large or small (close to zero) singular values
+    pub fn new(problem: Array2<A>, order: Order) -> TruncatedSvd<A> {
+        TruncatedSvd {
+            precision: 1e-5,
+            maxiter: problem.len_of(Axis(0)) * 2,
+            order,
+            problem,
+        }
+    }
+
+    /// Set the required precision of the solution
+    ///
+    /// The precision is, in the context of SVD, the square-root precision of the underlying
+    /// eigenproblem solution. The eigenproblem-precision is used to check the L2 error of each
+    /// eigenvector and stops its optimization when the required precision is reached.
+    pub fn precision(mut self, precision: f32) -> Self {
+        self.precision = precision;
+
+        self
+    }
+
+    /// Set the maximal number of iterations
+    ///
+    /// The LOBPCG is an iterative approach to eigenproblems and stops when this maximum
+    /// number of iterations are reached
+    pub fn maxiter(mut self, maxiter: usize) -> Self {
+        self.maxiter = maxiter;
+
+        self
+    }
+
+    /// Calculate the singular value decomposition
+    ///
+    /// # Parameters
+    ///
+    ///  * `num`: number of singular-value/vector pairs, ordered by magnitude
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use ndarray::{arr1, Array2};
+    /// use ndarray_linalg_rs::lobpcg::{TruncatedSvd, TruncatedOrder};
+    ///
+    /// let diag = arr1(&[1., 2., 3., 4., 5.]);
+    /// let a = Array2::from_diag(&diag);
+    ///
+    /// let eig = TruncatedSvd::new(a, TruncatedOrder::Largest)
+    ///    .precision(1e-5)
+    ///    .maxiter(500);
+    ///
+    /// let res = eig.decompose(3);
+    /// ```
+    pub fn decompose(self, num: usize) -> Result<TruncatedSvdResult<A>> {
+        if num < 1 {
+            panic!("The number of singular values to compute should be larger than zero!");
+        }
```

Comment:
Pass in `NonZeroUsize` instead of `usize`

Reply:
but then we need to wrap the decomposition size into a `NonZeroUsize` before calling? This adds much friction

Reply:
It's not unreasonable to ask users to provide an `NonZeroUsize` to a function when providing a 0 will result in a panic.

Reply:
I would just return an error, so that the user sees what happened

Reply:
I think we should just panic on 0, since that's the same behaviour as `Iterator::step_by` in the standard library. Not really any point in returning an error for something like this

Reply:
I'm just not sure if it's worth adding a new error variant just for this. Especially because the step size is difficult to get wrong, since it's usually specified as an integer literal. I'd prefer panicking on both. Also, why is using `NonZeroUsize` too much friction? It's the best of both worlds, since it can be constructed with or without panicking. The only drawback I can see is that it doesn't match the `decompose` API in `ndarray-linalg`, which uses `usize`.

Reply:
actually let's just return an empty solution when `num = 0`, I will remove the error variant and panic for zero step-size. I really looking forward when [constant constrains](https://nora.codes/post/its-time-to-get-hyped-about-const-generics-in-rust/) is hitting stable :/

> Also, why is using NonZeroUsize too much friction? 

I think that writing `svd.decompose(NonZeroUsize::new(5).unwrap());` takes too much space to encode the non-zeroness of eigenvalue number

Reply:
oh I just realized that we also have to bail out when `num > problem size` or return a solution with zero eigenvalues for excessive eigenvectors. I'll probably do the second